tsl/src/data_node.c

```diff
@@ -1320,13 +1320,140 @@ data_node_detach(PG_FUNCTION_ARGS)
 	PG_RETURN_INT32(removed);
 }
 
+/*
+ * Drop a data node's database.
+ *
+ * To drop the database on the data node, a connection must be made to another
+ * database since one cannot drop the database currently connected
+ * to. Therefore, we bypass the connection cache and use a "raw" connection to
+ * a standard database (e.g., template0 or postgres), similar to how
+ * bootstrapping does it.
+ *
+ * Note that no password is provided on the command line as is done when
+ * bootstrapping. Instead, it is assumed that the current user already has a
+ * method to authenticate with the remote data node (e.g., via a password
+ * file, certificate, or user mapping). This should normally be the case or
+ * otherwise the user wouldn't have been able to use the data node.
+ *
+ * Note that the user that deletes a data node also must be the database owner
+ * on the data node. The database will only be dropped if there are no other
+ * concurrent connections so all connections must be closed before being able
+ * to drop the database.
+ */
+static void
+drop_data_node_database(const ForeignServer *server)
+{
+	ListCell *lc;
+	TSConnection *conn;
+	Oid userid = GetUserId();
+	TSConnectionId connid = {
+		.server_id = server->serverid,
+		.user_id = userid,
+	};
+	/* Make a copy of the node name since the server pointer will be
+	 * updated */
+	char *nodename = pstrdup(server->servername);
+	char *dbname = NULL;
+	char *err = NULL;
+	int i;
+
+	/* Figure out the name of the database that should be dropped */
+	foreach (lc, server->options)
+	{
+		DefElem *d = lfirst(lc);
+
+		if (strcmp(d->defname, "dbname") == 0)
+		{
+			dbname = defGetString(d);
+			break;
+		}
+	}
+
+	if (NULL == dbname)
+	{
+		/* This should not happen unless the configuration is messed up */
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_DATA_NODE_INVALID_CONFIG),
+				 errmsg("could not drop the database on data node \"%s\"", nodename),
+				 errdetail("The data node configuration lacks the \"dbname\" option.")));
+		pg_unreachable();
+		return;
+	}
+
+	/* Clear potentially cached connection to the data node for the current
+	 * session so that it won't block dropping the database */
+	remote_connection_cache_remove(connid);
```

Comment:
Cached connections being dropped as well after the execution of this function in the `delete_data_node()`, maybe we should invalidate them first before executing `drop_data_node_database()` then?

Reply:
Invalidating a connection doesn't close it or remove it from the cache. It only marks it as invalid so that it is recreated on the next cache lookup.

We need to purge/close any of our own cached connections to the data node's database, or else it will block DROP DATABASE.  The purpose of this call is the close the connection and remove it from the cache. There's only one connection in the cache for each data node, so there's no point in removing other connections.